### PR TITLE
Added support for sections in tableview datasource

### PR DESCRIPTION
**Pull request by dehlen**
This adds the possibility to define an 2d-array of cell declarators in order to add support for multiple sections in the tableview datasource.

The following code snippet shows the usage:
```swift
public struct Data {
    var bar: Bar
    var baz: Baz

    var cellDeclarations: [[FooCellDeclaration]] {
        return [bar.cellDeclarations, baz.cellDeclarations]
    }
}

public struct Bar: CellsDeclarator {
    var bar: [BarCellConfiguration]

    public func declareCells(_ cell: (MyDmCellDeclaration) -> Void) {

        if !bar.isEmpty {
            for cellConfiguration in bar {
                cell(.someCell(cellConfiguration))
            }
        }
    }
} //same for Baz


//Data().cellDeclarations can be used as before
```

### CODE REVIEW

**Comment by ishkawa** [comment]
I need more explanation of the motivation of introducing sections to this library.

UITableView has sections actually, but I don't think this library should support sections. As Demo illustrates, we already have way to display section headers, so we need strong motivation if we introduce the second way to do the same thing.

**Comment by dehlen** [comment]
I am currently at the FrenchKit conference but I‘ll come back to your question later this day.

**Comment by dehlen** [comment]
Basically the motivation for this change is related to the way UITableView handles separators. If the section headers are just other cell types the section header will get a separator aswell. Of course you could hack around that by setting seperatorInsets or adding a custom seperator to all other cell types and at the same time disable separators for UITableView but I think supporting sections headers in the datasource is the most correct solution to this problem. Also with the section headers support you gain extra functionality like sticky section headers (like in the contact iOS app for example).

**Comment by dehlen** [comment]
I updated the PR to resolve the merge conflicts. What are your thoughts on my comment above ? I just need to know whether I should maintain a fork or if you think this might be a valuable addition to the framework.

**Comment by ishkawa** [comment]
I understood the motivation. It is nice to implement section headers in canonical way. However, I still have 2 following concerns.

- Definition of contents is separated into 2 places.
- Ways to define contents are different each other. Contents of each sections are defined in declarative way, but collection of them are defined by array. 

If we support sections in CellsDeclarator, I think everything should be defined in declarative way like below. 

```swift
struct Data: CellsDeclarator {
    var venue: Venue
    var reviews: [Review]
    var relatedVenues: [Venue]
    
    func declareCells(
        sectionHeader: (SectionHeaderDeclaration) -> Void,
        cell: (CellDeclaration) -> Void,
        sectionFooter: (SectionFooterDeclaration) -> Void) {

        cell(.outline(venue))

        if !reviews.isEmpty {
            sectionHeader("Reviews")
            for review in reviews {
                cell(.review(review))
            }
        }
        
        if !relatedVenues.isEmpty {
            sectionHeader("Related Venues")
            for relatedVenue in relatedVenues {
                cell(.relatedVenue(relatedVenue))
            }
        }
    }
}
```

Unfortunately,  this code introduces another problem because it is not appropriate for some of UICollectionViewLayout.

In short, I'm positive to support sections, but haven't found a reasonable way at this moment.

**Comment by ishkawa** [comment]
[As mentioned in another pull request](https://github.com/ishkawa/DataSourceKit/pull/10#issuecomment-428631250), we might support sections by type associated with CellBinder. I will share this idea later.

**Comment by dehlen** [comment]
Did you find time to iterate on your idea ? I would love to get an update on this one.
